packages/firestore/test/unit/remote/bloom_filter.test.ts

```diff
@@ -0,0 +1,248 @@
+      3,
+      13
+    );
+    expect(bloomFilter.getSize()).to.equal(37);
+  });
+
+  it('should throw error if padding is invalid', () => {
+    try {
+      new BloomFilter(new Uint8Array(1), -1, 1);
+      expect.fail();
+    } catch (error) {
+      expect(
+        (error as Error)?.message.includes(
+          'INTERNAL ASSERTION FAILED: Invalid padding: -1'
+        )
+      ).to.be.true;
+    }
+    try {
+      new BloomFilter(new Uint8Array(1), 9, 1);
+      expect.fail();
+    } catch (error) {
+      expect(
+        (error as Error)?.message.includes(
+          'INTERNAL ASSERTION FAILED: Invalid padding: 9'
+        )
+      ).to.be.true;
+    }
+  });
+
+  it('should throw error if hash count is negative', () => {
+    try {
+      new BloomFilter(new Uint8Array(1), 1, -1);
+      expect.fail();
+    } catch (error) {
+      expect(
+        (error as Error)?.message.includes(
+          'INTERNAL ASSERTION FAILED: Invalid hash count: -1'
+        )
+      ).to.be.true;
+    }
+  });
+
+  it('should throw error if hash count is 0 for non empty bloom filter', () => {
+    try {
+      new BloomFilter(new Uint8Array(1), 1, 0);
+      expect.fail();
+    } catch (error) {
+      expect(
+        (error as Error)?.message.includes(
+          'INTERNAL ASSERTION FAILED: Invalid hash count: 0'
+        )
+      ).to.be.true;
+    }
+  });
+
+  it('mightContain in empty bloom filter should always return false', () => {
+    const bloomFilter = new BloomFilter(new Uint8Array(0), 0, 0);
+    expect(bloomFilter.mightContain('abc')).to.be.false;
+    expect(bloomFilter.mightContain('def')).to.be.false;
+  });
+
+  it('mightContain should always return false for empty string', () => {
+    const emptyBloomFilter = new BloomFilter(new Uint8Array(0), 0, 0);
+    const nonEmptyBloomFilter = new BloomFilter(
+      new Uint8Array([70, 204, 25]),
+      1,
+      16
+    );
+    expect(emptyBloomFilter.mightContain('')).to.be.false;
+    expect(nonEmptyBloomFilter.mightContain('')).to.be.false;
+  });
+
+  /**
+   * Golden tests are generated by backend based on inserting n number of
+   * documents path into a bloom filter.
+   *
+   * Full document path is generated by concating documentPrefix and number n,
+   * ie, projects/project-1/databases/database-1/documents/coll/doc12
+   *
+   * The test result is generated by checking the membership of documents from
+   * documentPrefix+0 to documentPrefix+2n. The membership results from 0 to n
+   * is expected to be true (1), and the membership results from n to 2n is
+   * expected to be false (0) with some false positive results (1).
+   *
+   * The test result of BloomFilter.mightContain() should match the backend
+   * result exactly.
+   */
+  describe('BloomFilter membership test', () => {
+    const documentPrefix =
+      'projects/project-1/databases/database-1/documents/coll/doc';
+
+    interface TestDataType {
+      bits: {
+        bitmap?: string;
```

Comment:
Do all of these properties need to be optional? i.e. `bitmap?`, `padding?`, and `hashCount?`

Shouldn't these properties _always_ be present?

Reply:
Are these optional because some of the test files omit these values? If so, let's get those value to be explicitly specified in the golden test files. It's best to be as explicit as possible in tests for maximum readability. Every piece of "conditional" logic in tests makes them more difficult to reason about, and humans' ability to reason about tests is the only way that the correctness of the tests is validated.